### PR TITLE
Contacts Table Pagination

### LINKED ISSUE #24: [ENG] Support pagination to contact page
# Context
Right now, we only display the 50 first results. We don't have any pagination logic :( 
Given that it will take a while for orgs to onboard more than 50 volunteers / send more than 50 newsletters/ for a contact to receive more than 50 letters, we can just focus on adding pagination to the contact page (which will very quickly have more than 50 contacts)

# To-Do
- When users scroll to the bottom, we want to issue another request to fetch the next page
- You'll probably need to add a page number to the redux store
- Once users hit the bottom of the page, we want to show a spinner and issue the fetch request, passing in the page we are requesting

### PR DESCRIPTION
**Pull request by Sars47**
## Motivation and Context

Closes #24 -- however this PR needs to be followed up with a new PR that generalizes the pagination logic for reuse across contacts, volunteers, and letters.

## How Has This Been Tested?

Tested in Brave & Chrome up to 850 contacts.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

### CODE REVIEW

File: src/components/pages/Contacts/index.tsx [hunk 1]
```diff
@@ -117,6 +116,20 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
     page_id = 'faded';
   }
 
+  const handleScroll = (e: UIEvent) => {
+    const tableDiv = document.getElementById('tableDiv');
+    // const org = user.user.org;
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:121]: should we delete this?

File: src/components/pages/Contacts/index.tsx [hunk 2]
```diff
@@ -117,6 +116,20 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
...
+
+    if (tableDiv) {
+      const isBottom =
+        tableDiv.scrollHeight - tableDiv.scrollTop === tableDiv.clientHeight;
+      if (isBottom && org) {
+        console.log('table bottom reached');
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:127]: guess we can delete this one as well!

File: src/components/pages/Contacts/index.tsx [hunk 3]
```diff
@@ -55,16 +54,16 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
   );
   const [hasFetchedData, setHasFetchedData] = useState<boolean>(false);
 
+  const contacts = orgContacts.contacts;
   const token = user.user.token;
   const org = user.user.org;
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:59]: nit: should we just destruct to be more ES6-y?
const { contacts}  = orgContacts;
const {token} = user.user;
const {org} = user.user

File: src/components/pages/Contacts/index.tsx [hunk 4]
```diff
@@ -117,6 +116,20 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
     page_id = 'faded';
   }
 
+  const handleScroll = (e: UIEvent) => {
+    const tableDiv = document.getElementById('tableDiv');
+    // const org = user.user.org;
+
+    if (tableDiv) {
+      const isBottom =
+        tableDiv.scrollHeight - tableDiv.scrollTop === tableDiv.clientHeight;
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:125]: woww, this is awesome!!

File: src/services/Api/contacts.ts [hunk 5]
```diff
@@ -15,7 +16,7 @@ export async function fetchContacts(
     },
   };
   const response = await fetch(
-    url.resolve(API_URL, `contacts/org/${org_id}`),
+    url.resolve(API_URL, `contacts/org/${org_id}?page=${page}`),
```
Comment (gabrielsaruhashi) [inline src/services/Api/contacts.ts:19]: 💯

File: src/redux/modules/orgcontacts.ts [hunk 6]
```diff
@@ -203,7 +204,12 @@ export function orgContactsReducer(
 ): OrgContactsState {
   switch (action.type) {
     case SET_ORG_CONTACTS:
-      return { ...state, contacts: action.payload, loading: false };
+      return {
+        ...state,
+        contacts: state.contacts.concat(action.payload),
```
Comment (gabrielsaruhashi) [inline src/redux/modules/orgcontacts.ts:209]: love to see it!! this is awesomeeeee 🔥

File: src/components/pages/Contacts/index.tsx (hunk 4 above)
Comment (gabrielsaruhashi) [reply src/components/pages/Contacts/index.tsx:125]: do we display any feedback to the user while the bottom is loading?

File: src/components/pages/Contacts/index.tsx [hunk 7]
```diff
@@ -117,6 +116,20 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
     page_id = 'faded';
   }
 
+  const handleScroll = (e: UIEvent) => {
+    const tableDiv = document.getElementById('tableDiv');
+    // const org = user.user.org;
+
+    if (tableDiv) {
+      const isBottom =
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:124]: maybe we can create a isBottom function on the util file? Seems like a useful function to have around!

File: src/components/pages/Contacts/index.tsx (hunk 4 above)
Comment (gabrielsaruhashi) [reply src/components/pages/Contacts/index.tsx:125]: maybe we can show spinner while it loads if we're not!

Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:125]: Right now I have it showing the wheel as it did when loading all the contacts before

Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:125]: But this does cover all the contacts while the new contacts are loading (which in prod could be a non-trivial amounts of time) so maybe we want to change this so the wheel displays below the contacts that have already been loaded

Comment (gabrielsaruhashi) [reply src/components/pages/Contacts/index.tsx:125]: oh so it just renders on top of the table?
ye! I think generalizing it to render below contacts makes sense. But we can do it as a fast follow if needed

I am also curious how we should handle contact searches.
Like if someone looks up a contact that's not being rendered it just doesn't show up. 
We might need to create an endpoint in the future to search for specific contacts and concat them to the array of contacts

Will open issue for this!

File: src/components/pages/Contacts/index.tsx (hunk 1 above)
Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:121]: Deleted!

File: src/components/pages/Contacts/index.tsx (hunk 2 above)
Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:127]: Deleted!

File: src/components/pages/Contacts/index.tsx (hunk 3 above)
Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:59]: that is new syntax to me - but i'm a big fan. Just switched it

File: src/components/pages/Contacts/index.tsx (hunk 4 above)
Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:125]: yeah that makes a lot of sense - We can do a fast follow PR for loading wheel at bottom - maybe baked into a PR to generalize the logic for the other 2 identified uses?

File: src/components/pages/Contacts/index.tsx (hunk 7 above)
Comment (Sars47) [reply src/components/pages/Contacts/index.tsx:124]: yeah that sounds good - put into utils

File: src/components/pages/Contacts/index.tsx [hunk 8]
```diff
@@ -117,6 +116,16 @@ const UnconnectedContacts: React.FC<PropsFromRedux> = ({
     page_id = 'faded';
   }
 
+  const handleScroll = (e: UIEvent) => {
+    const tableDiv = document.getElementById('tableDiv');
+
+    if (tableDiv) {
+      if (isBottom(tableDiv) && org) {
```
Comment (gabrielsaruhashi) [inline src/components/pages/Contacts/index.tsx:123]: oh is org nullable ? 
we should probably change this interface in the future. Everyone that reaches this page should be affiliated with an org

**Review by gabrielsaruhashi — DISMISSED** [review]
@Sars47 kicking some ass
alt text

**Comment by gabrielsaruhashi** [comment]
u probably just wanna check out this conflict!

**Pull request merged**